Add caching for data permissions

johnswanson opened the pull request:
This caching is:

- thread-local. It's intended to be used just like
`api/*current-user-permission-set*` (although less directly): we use
middleware to wrap a single request/response handling cycle in
`with-relevant-permissions-for-user` to populate the cache, and then the
cache is used.

- scoped to a particular user. It will not be used if the
`api/*current-user-id*` isn't equal to the `user-id` in a particular
permissions query. The `user-id` is also used as an index when
retrieving permissions to prevent any possibility of accidentally using
cached permissions for a different user than intended.

- "indexed" by `[user-id, perm-type, db-id]` - the former for the above
reason, the latter two (perm-type and db-id) because essentially every
time we need to retrieve permissions we filter by these two columns.

To avoid having multiple codepaths for filtering permissions (e.g. one for cached vs. another for uncached), for now I'm retrieving all of a user's relevant permissions in *both* cases, then filtering them down when we calculate a given permission. Less efficient, yes, but this way we are always testing the same logic, where otherwise we might have subtle bugs where we filtered permissions differently in SQL vs from the cache. And I don't think we *ever* calculate permissions in production that *aren't* for the currently logged in user, so this shouldn't have any negative performance impact in prod.

(Edit: I actually decided that instead of select all of the user's relevant permissions, we could just select the ones related to the perm_type and db_id we're interested in.)

Closes [38040](https://github.com/metabase/metabase/issues/38040)

johnswanson committed:
Add caching for data permissions

This caching is:

- thread-local. It's intended to be used just like
`api/*current-user-permission-set*` (although less directly): we use
middleware to wrap a single request/response handling cycle in
`with-relevant-permissions-for-user` to populate the cache, and then the
cache is used.

- scoped to a particular user. It will not be used if the
`api/*current-user-id*` isn't equal to the `user-id` in a particular
permissions query. The `user-id` is also used as an index when
retrieving permissions to prevent any possibility of accidentally using
cached permissions for a different user than intended.

- "indexed" by `[user-id, perm-type, db-id]` - the former for the above
reason, the latter two (perm-type and db-id) because essentially every
time we need to retrieve permissions we filter by these two columns.

johnswanson committed:
Use `relevant-permissions-for-user` for all perms

I had one point of slight uneasiness with my previous implementation:
each permissions check had two branches (cached and uncached). For an
example on the more complicated end of the specturm, when checking the
"schema" permission for a user, we did:

```
(if (use-cached-values? user-id)
  (->> (get-cached-permissions user-id perm-type database-id)
       (filter #(or (= (:schema_name %) schema-name)
                    (nil? (:table_id %))))
       (map #(set/rename-keys % {:group_id :group-id
                                 :perm_value :value}))
       (map #(select-keys % [:group-id :value])))
  (t2/select :model/DataPermissions
             {:select [[:p.group_id :group-id]
                       [:p.perm_value :value]]
              :from [[:permissions_group_membership :pgm]]
              :join [[:permissions_group :pg] [:= :pg.id :pgm.group_id]
                     [:data_permissions :p]   [:= :p.group_id :pg.id]]
              :where [:and
                      [:= :pgm.user_id user-id]
                      [:= :p.perm_type (u/qualified-name perm-type)]
                      [:= :p.db_id database-id]
                      [:or
                       [:= :schema_name schema-name]
                       [:= :table_id nil]]]}))
```

For each kind of permission check, we need to duplicate this kind of
select logic across two codepaths. Then, we'll *almost always* be using
the cached codepath in production. This made me... a little uneasy?

So instead, it seems safer to *always* retrieve all relevant permissions
for a given user and then filter them the same way in every case, rather
than having two code paths.

This is slightly less efficient, of course: we're getting more data then
we need from the database and then sifting through it locally. One easy
optimization would be to optionally filter by `perm_type` and `db_id` in
`relevant-permissions-for-user`. But - in production, I'm not sure
we *ever* actually check any user's permissions besides the currently
logged in user. So any optimization here seems like probably
over-optimization.

johnswanson committed:
slight optimization for noncached permissions

Select slightly fewer rows for local filtering for noncached permissions.

johnswanson committed:
oops - bugfix

noahmoss commented on test/metabase/models/data_permissions_test.clj:
```diff
@@ -151,6 +165,7 @@
                  :model/Database                   {database-id :id} {}
                  :model/Table                      {table-id-1 :id}  {:db_id database-id}
                  :model/Table                      {table-id-2 :id}  {:db_id database-id}]
+    ;; for cache
```
don't understand this comment

johnswanson replied:
Doh, sorry - I started writing a comment, then moved things around and forgot to delete it.

noahmoss commented on src/metabase/models/data_permissions.clj:
```diff
@@ -277,17 +292,9 @@
-                                [:= :pgm.user_id user-id]
-                                [:= :p.perm_type (u/qualified-name perm-type)]
-                                [:= :p.db_id database-id]]})]
+    (let [perm-values (->> (get-permissions user-id perm-type database-id)
+                             (map :perm_value)
+                             (into #{}))]
```
```suggestion
    (let [perm-values (->> (get-permissions user-id perm-type database-id)
                           (map :perm_value)
                           (into #{}))]
```

noahmoss reviewed (approved):
I like this approach a lot! Refactoring out the SQL query for fetching perms is a very nice benefit too. Just a couple nits.

I guess the only remaining component is to hook this up to the session middleware? Were you planning on doing that in a separate PR?

noahmoss commented on test/metabase/models/data_permissions_test.clj:
```diff
@@ -138,7 +139,20 @@
 
       (testing "Admins always have the most permissive value, regardless of group membership"
-        (is (= :yes (data-perms/database-permission-for-user (mt/user->id :crowberto) :perms/native-query-editing database-id-2)))))))
+        (is (= :yes (data-perms/database-permission-for-user (mt/user->id :crowberto) :perms/native-query-editing database-id-2)))))
+    (testing "caching works as expected"
```
```suggestion
        (is (= :yes (data-perms/database-permission-for-user (mt/user->id :crowberto) :perms/native-query-editing database-id-2)))))
        
    (testing "caching works as expected"
```

noahmoss commented on test/metabase/models/data_permissions_test.clj:
```diff
@@ -162,7 +177,20 @@
 
       (testing "Admins always have the most permissive value, regardless of group membership"
-        (is (= :unrestricted (data-perms/table-permission-for-user (mt/user->id :crowberto) :perms/data-access database-id table-id-2)))))))
+        (is (= :unrestricted (data-perms/table-permission-for-user (mt/user->id :crowberto) :perms/data-access database-id table-id-2)))))
+      (mt/with-restored-data-perms-for-groups! [group-id-1 group-id-2]
```
```suggestion
        (is (= :unrestricted (data-perms/table-permission-for-user (mt/user->id :crowberto) :perms/data-access database-id table-id-2)))))
        
     (mt/with-restored-data-perms-for-groups! [group-id-1 group-id-2]
```

noahmoss replied:
(looks like this was at the same indentation as the previous `testing` block, but the parens weren't balanced, or something...not sure exactly but you might want to check the indentation here)

johnswanson commented:
> I guess the only remaining component is to hook this up to the session middleware? Were you planning on doing that in a separate PR?

That was my plan, but actually, I might as well do it here - I'll add that.

johnswanson committed:
check if the cache exists, not if there are rows

With the previous implementation, imagine we get relevant permissions
for a `nil` user-id (e.g. when embedding).

`(group-by (juxt :user_id :perm_type :db_id) [])` is going to return an
empty map.

Then `(get @*permissions-for-user* [user-id perm-type db-id])` will
return `nil`, and we'll make a database query to get the relevant
permissions for the (`nil`) user again unnecessarily.

Instead, we store the cache as a nested map. At the lowest level, the
keys are user-ids. This way, we can check to see whether we've retrieved
the cache for a particular user. If we have, we don't need to do it again.

johnswanson committed:
Add data-perms cache to session middleware

johnswanson commented:
@noahmoss I added the data-perms cache to the session middleware in the latest commit - quick and easy 🎉 

It also made me think about a potential flaw in my implementation of the cache - specifically, if we don't get _any_ rows back for a given user ID (as might happen for a `nil` user id) we'd always hit the DB, because `(group-by (juxt :user_id :perm_type :db_id) [])` will return `{}`, and `(get {} [user-id perm-type db-id])` will return `nil`.

I modified things slightly to instead have a nested map with user-id keys. Then, to check whether the cache exists, we can just do `(get {} user-id)`, which will return a truthy value (`{}`) if we've cached the relevant rows (i.e. no rows) already. Does that seem reasonable?

noahmoss commented:
@johnswanson Sounds good!
